component/http/handler.go

```diff
@@ -62,44 +62,76 @@ func determineEncoding(r *http.Request) (string, encoding.DecodeFunc, encoding.E
 	var enc encoding.EncodeFunc
 	var dec encoding.DecodeFunc
 	var ct string
+	var foundValidHeader bool
 
 	if cok {
-		switch cth[0] {
-		case "*/*", json.Type, json.TypeCharset:
-			enc = json.Encode
-			dec = json.Decode
-			ct = json.TypeCharset
-		case protobuf.Type, protobuf.TypeGoogle:
-			enc = protobuf.Encode
-			dec = protobuf.Decode
-			ct = protobuf.Type
-		default:
+		foundValidHeader = false
+		ah := getMultiValueHeaders(cth[0])
+		for _, v := range ah {
+			switch v {
+			case "*/*", json.Type, json.TypeCharset:
+				enc = json.Encode
+				dec = json.Decode
+				ct = json.TypeCharset
+				foundValidHeader = true
+			case protobuf.Type, protobuf.TypeGoogle:
+				enc = protobuf.Encode
+				dec = protobuf.Decode
+				ct = protobuf.Type
+				foundValidHeader = true
+			}
+		}
+		// No valid headers found after going through all headers
+		if !foundValidHeader {
 			return "", nil, nil, errors.New("content type header not supported")
 		}
 	}
 
 	if aok {
-		switch ach[0] {
-		case "*/*", json.Type, json.TypeCharset:
-			enc = json.Encode
-			if dec == nil {
-				dec = json.Decode
-			}
-			ct = json.TypeCharset
-		case protobuf.Type, protobuf.TypeGoogle:
-			enc = protobuf.Encode
-			if dec == nil {
-				dec = protobuf.Decode
+		foundValidHeader = false
+		ah := getMultiValueHeaders(ach[0])
+		for _, v := range ah {
+			switch v {
+			case "*/*", json.Type, json.TypeCharset:
```

Comment:
With the above for loop the last match wins. Maybe we should break when the first match happens.